File: server/datasets/python_source-code/Multiple-D1/organic_6.py
"""
Quick Sort - Organic Submission #6
Three-way partitioning for arrays with high duplicates.
"""

def three_way_partition(arr, low, high):
    pivot = arr[low]
    lt = low
    gt = high
    i = low + 1
    while i <= gt:
        if arr[i] < pivot:
            arr[lt], arr[i] = arr[i], arr[lt]
            lt += 1
            i += 1
        elif arr[i] > pivot:
            arr[gt], arr[i] = arr[i], arr[gt]
            gt -= 1
        else:
            i += 1
    return lt, gt

def quick_sort_3way(arr, low, high):
    if low < high:
        lt, gt = three_way_partition(arr, low, high)
        quick_sort_3way(arr, low, lt - 1)
        quick_sort_3way(arr, gt + 1, high)

if __name__ == "__main__":
    nums = [4, 2, 4, 4, 1, 3, 2, 4, 1]
    quick_sort_3way(nums, 0, len(nums) - 1)
    print(nums)

File: server/datasets/java_source-code/Multiple-D1/organic_6.java
// Organic Student Submission 6
public class DualPivotQuickSort {
    public static void dualPivotSort(int[] arr, int low, int high) {
        if (low < high) {
            if (arr[low] > arr[high]) {
                int temp = arr[low]; arr[low] = arr[high]; arr[high] = temp;
            }
            int p = arr[low], q = arr[high];
            int l = low + 1, g = high - 1, k = l;
            while (k <= g) {
                if (arr[k] < p) {
                    int t = arr[k]; arr[k] = arr[l]; arr[l] = t;
                    l++;
                } else if (arr[k] >= q) {
                    while (arr[g] > q && k < g) g--;
                    int t = arr[k]; arr[k] = arr[g]; arr[g] = t;
                    g--;
                    if (arr[k] < p) {
                        int t2 = arr[k]; arr[k] = arr[l]; arr[l] = t2;
                        l++;
                    }
                }
                k++;
            }
            l--; g++;
            int t1 = arr[low]; arr[low] = arr[l]; arr[l] = t1;
            int t2 = arr[high]; arr[high] = arr[g]; arr[g] = t2;
            dualPivotSort(arr, low, l - 1);
            dualPivotSort(arr, l + 1, g - 1);
            dualPivotSort(arr, g + 1, high);
        }
    }
}
